src/profile-logic/transforms.js

```diff
@@ -906,70 +906,87 @@ export function collapseDirectRecursion(
   funcToCollapse: IndexIntoFuncTable,
   implementation: ImplementationFilter
 ): Thread {
+  // Collapse recursion by reparenting stack nodes for all "inner" frames of a
+  // recursion to the same level as the outermost frame.
+  //
+  // Example with recursion on B:
+  //  - A1                    - A1
+  //    - B1                    - B1
+  //      - B2                  - B2
+  //        - B3        ->      - B3
+  //           - C1               - C1
+  //      - B4                  - B4
+  //        - C2                  - C2
+  //
+  // In the call tree, sibling stack nodes with the same function will be
+  // collapsed into one call node.
+  // We keep all the stack nodes and frames, we just rewire them such that the
+  // outer stack nodes of the recursion are skipped. We skip the outer nodes
+  // rather than the inner nodes, so that per-frame data such as line numbers and
+  // frame addresses are counted for the innermost frame in a stack. We prefer
+  // keeping this information for the innermost frame because the outer frames
+  // just have the line and instruction address of the recursive call, and the
+  // purpose of "collapsing recursion" is to ignore that recursive call.
+  //
+  // Applying the transform's implementation filter is done on a best effort
+  // basis and doesn't really have a clean solution. We want to make the
+  // following case work:
+  // Full:     Ajs -> Xcpp -> Bjs -> Ycpp -> Bjs -> Zcpp -> Bjs -> Wcpp
+  // JS-only:  Ajs -> Bjs -> Bjs -> Bjs
+  // Now collapse recursion on Bjs.
+  // Collapsed JS-only:  Ajs -> Bjs
+  // Now switch back to all stack types.
+  // Collapsed full:     Ajs -> Xcpp -> Bjs -> Wcpp
+
   const { stackTable, frameTable } = thread;
-  const oldStackToNewStack: Map<
-    IndexIntoStackTable | null,
-    IndexIntoStackTable | null
-  > = new Map();
-  // A root stack's prefix will be null. Maintain that relationship from old to new
-  // stacks by mapping from null to null.
-  oldStackToNewStack.set(null, null);
-  const recursiveStacks = new Set();
-  const newStackTable = getEmptyStackTable();
+  const recursionChainPrefixForStack = new Map();
   const funcMatchesImplementation = FUNC_MATCHES[implementation];
+  const newStackTablePrefixColumn = stackTable.prefix.slice();
```

Comment:
Interesting. It looks like previously we were updating the whole stack table and samples table, but now only updating the prefix column of the stack table. This is because we don't need to touch the samples table anymore since we are always keeping the leaf-most stack right? This looks faster also!

Reply:
Yes exactly, this implementation has the advantage that the stack indexes don't change, so none of the other tables need to be updated with new stack indexes.